clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -222,6 +223,27 @@ public void clearPartitionRecords() {
         partitionSizes();
         data = null;
     }
+    
+    public static void validateProducerIds(short version, ProduceRequestData data) {
+        if (version >= 3) {
+            long producerId = -1;
+            for (ProduceRequestData.TopicProduceData topicData : data.topicData()) {
+                for (ProduceRequestData.PartitionProduceData partitionData : topicData.partitionData()) {
+                    BaseRecords baseRecords = partitionData.records();
+                    if (baseRecords instanceof Records) {
+                        Records records = (Records) baseRecords;
+                        for (RecordBatch batch : records.batches()) {
+                            if (producerId == -1 && batch.hasProducerId())
```

Comment:
Is it an error if there's one batch with producer ID and another without one?

Reply:
Based on some of the ProduceRequest tests I found, we allow for some batches with and some without. 
See `testMixedIdempotentData`